### PR TITLE
Enhancing the possibilities of the class attribute

### PR DESCRIPTION
**Pull request by Yomguithereal**
Proposition to enhance the possibilities of the `class` attribute of the virtual elements. This is inspired by the full possibilities of [this](https://www.npmjs.com/package/classnames) popular package.

You can now do the following:

```jsx
// Compact array (dropping falsey values)
<div class={['foo', false, 'baz']} />

// Array of mixed strings and objects
<div class={['foo', {bar: true}, 'baz']} />
```

I've used a succession of `map`, `filter` and `join` in the code. Whether you'd accept the idea, just tell me whether you want me to low-level this for perfs.

Sorry for the PR flood again :-).

### CODE REVIEW

**Comment by anthonyshort** [comment]
I was thinking about pulling this out of the virtual element function for now so that there is less magic. It would be possible to just wrap the `element` function in your own function that adds some magical attributes like this though so you didn't need to do it every time. 

PR flood is awesome! Keep it up :D

**Comment by anthonyshort** [comment]
Pulling out all of the magical attributes means that we get some perf boosts because the virtual nodes will be quicker to create rather than wasting time iterating over all the attributes and child nodes just to create the virtual element. 

Here's the project for the new virtual-element module pulled out of Deku: https://github.com/dekujs/virtual-element

**Comment by Yomguithereal** [comment]
Ok. Do you have an example somewhere on how you think you'll perform the virtual nodes "enhancement"?

**Comment by anthonyshort** [comment]
Sure! You could do something like this:

```js
import classnames from 'classnames'
import element from 'virtual-element'

export default function enhancedElement (name, attrs, children) {
  for (var attr in attrs) {
    if (attr === 'class') attrs[attr] = classnames(attrs[attr])
  }
  return element(name, attrs, children)
}
```

**Comment by Yomguithereal** [comment]
And so you'd be able to tell deku to use our enhanced element app-wide with an option, for instance?

**Comment by anthonyshort** [comment]
In the next version you won't need to. I've started working on the readme for this: https://github.com/dekujs/deku/blob/edge/README.md

tldr; as long as the component returns an object of a certain format, it'll render anything.

**Comment by Yomguithereal** [comment]
Cool, so by changing `jsxPragma` we can use anything in JSX too I guess. This seems very cool!

**Comment by Yomguithereal** [comment]
Closing this PR as it is obsolete compared to edge. By the way, do you have an idea of when should the edge version released?

**Comment by anthonyshort** [comment]
It shouldn't be too long. I just haven't had a good chunk of time to work on it lately. We really want to get it out though, since we need it internally :)

**Comment by joshrtay** [comment]
@anthonyshort In user land what will the best practices be for packaging "enhanced virtual elements" with components that depend on them. Before you could just export the component object for deku component module, but now there seems to be coupling between "enhanced virtual elements" and components.

**Comment by Yomguithereal** [comment]
Would you qualify the stability of the current edge version right now? I need to start something and won't be affected if some things are amiss and should be able to report them if needed.

**Comment by anthonyshort** [comment]
@joshrtay There will always be some coupling between the virtual elements and the components, but even these wrapper libs would just be pure functions that return objects. Components can depend on any virtual dom library they want, as long as they conform to the contract of returning objects that look like this:

```
{
  type: Any
  attributes: { key: value }
  children: []
}
```

You can do whatever you want as long as the render function returns something that, at the very least, has these 3 properties.

**Comment by anthonyshort** [comment]
@Yomguithereal It's not ready for usage yet, a lot of the tests don't pass yet :)

**Comment by anthonyshort** [comment]
I'm aiming for this week for a release though.
